### PR TITLE
fetching gas prices from bundler and imp

### PR DESCRIPTION
**Pull request by talhamalik883**
This PR includes following changes
1. Gas prices fetching shifted from provider to bundler
2. custom verificationGasLimit is added in sendGaslessTransaction function to make it dynamic so that dapp can pass
3. custom verificationGasLimit and gasLimit is added in **sendGaslessTransactionBatch**, **updateImplementation**, **updateFallbackHandler**, **deployWalletUsingPaymaster**
4. Added Gas estimation bypassing flag in estimation endpoint for acme

### CODE REVIEW

File: packages/account-abstraction/src/ERC4337EthersSigner.ts [hunk 1]
```diff
@@ -30,7 +30,8 @@ export class ERC4337EthersSigner extends Signer {
   async sendTransaction(
     transaction: TransactionRequest,
     walletDeployOnly = false,
-    engine?: any // EventEmitter
+    vGasLimit?: BigNumber,
+    engine?: any, // EventEmitter
   ): Promise<TransactionResponse> {
```
Comment (AmanRaj1608) [inline packages/account-abstraction/src/ERC4337EthersSigner.ts:35]: Is there any EventEmitter type from node which we can assign here

File: packages/node-client/src/NodeClient.ts [hunk 2]
```diff
@@ -165,6 +165,10 @@ class NodeClient implements INodeClient {
   estimateHandlePaymentGas(
     estimateHandlePaymentTxGasDto: EstimateHandlePaymentTxGasDto
   ): Promise<EstimateGasResponse> {
+    const byPass = process.env.BY_PASS_ESTIMATION
+    if ( byPass === 'true' ){
+      estimateHandlePaymentTxGasDto.byPassEstimation = true
```
Comment (AmanRaj1608) [inline packages/node-client/src/NodeClient.ts:170]: Should check for both string and boolean. Can create a helper method for this in Nodeclient only

File: packages/account-abstraction/src/ERC4337EthersSigner.ts [hunk 3]
```diff
@@ -59,15 +60,15 @@ export class ERC4337EthersSigner extends Signer {
     }
 
     delete transaction.customData
-
+    
     let userOperation: UserOperation
     if (walletDeployOnly === true) {
```
Comment (AmanRaj1608) [inline packages/account-abstraction/src/ERC4337EthersSigner.ts:65]: Can run lint command globally.

**Review by AmanRaj1608 — COMMENTED** [review]
Looks good, so this PR has changes for 
- gasPrices fix by bundler
- bypass handle payment estimation

should we merge to dev first?

File: packages/account-abstraction/src/BaseAccountAPI.ts [hunk 4]
```diff
@@ -124,7 +124,7 @@ export abstract class BaseAccountAPI {
    * @param info
    */
   // UserOperation = UserOperationStruct // if anything changes changes in core-types
-  abstract createUnsignedUserOp(info: TransactionDetailsForBatchUserOp): Promise<UserOperation>
+  abstract createUnsignedUserOp(info: TransactionDetailsForBatchUserOp, vGasLimit?: BigNumber): Promise<UserOperation>
```
Comment (livingrockrises) [inline packages/account-abstraction/src/BaseAccountAPI.ts:127]: what is vGasLimit?
can we make variable names more understandable

File: packages/account-abstraction/src/BaseAccountAPI.ts [hunk 5]
```diff
@@ -235,7 +235,7 @@ export abstract class BaseAccountAPI {
       ) {
         return {
           callData: '0x',
-          callGasLimit: BigNumber.from('21000')
+          callGasLimit: BigNumber.from(detailsForUserOp?.gasLimit) ?? BigNumber.from(21000)
```
Comment (livingrockrises) [inline packages/account-abstraction/src/BaseAccountAPI.ts:238]: what is the purpose of this change?

File: packages/account-abstraction/src/BaseAccountAPI.ts [hunk 6]
```diff
@@ -356,8 +356,8 @@ export abstract class BaseAccountAPI {
    * helper method: create and sign a user operation.
    * @param info transaction details for the userOp
    */
-  async createSignedUserOp(info: TransactionDetailsForBatchUserOp): Promise<UserOperation> {
-    return await this.signUserOp(await this.createUnsignedUserOp(info))
+  async createSignedUserOp(info: TransactionDetailsForBatchUserOp, vGasLimit?: BigNumber): Promise<UserOperation> {
```
Comment (livingrockrises) [inline packages/account-abstraction/src/BaseAccountAPI.ts:359]: this gas limit is relevant with fetching gas price from bundler api how?

File: packages/account-abstraction/src/ERC4337EthersSigner.ts [hunk 7]
```diff
@@ -59,15 +60,15 @@ export class ERC4337EthersSigner extends Signer {
...
       userOperation = await this.smartAccountAPI.createSignedUserOp({
         target: [''],
         data: [''],
         value: [0],
-        gasLimit: [21000]
-      })
+        gasLimit: transaction.gasLimit
```
Comment (livingrockrises) [inline packages/account-abstraction/src/ERC4337EthersSigner.ts:70]: reason fro changing from 21k here?

File: packages/account-abstraction/src/Provider.ts [hunk 8]
```diff
@@ -22,7 +22,14 @@ export async function newProvider(
   const entryPoint = EntryPoint__factory.connect(config.entryPointAddress, originalProvider)
   // Initial SimpleWallet instance is not deployed and exists just for the interface
   // const simpleWalletDeployer = await DeterministicDeployer.deploy(SimpleAccountDeployer__factory.bytecode)
+  const httpRpcClient = new HttpRpcClient(
```
Comment (livingrockrises) [inline packages/account-abstraction/src/Provider.ts:25]: eventually initialising bundler client won't be tied with needing to pass EP address and DappAPIKey etc..

File: packages/account-abstraction/src/SmartAccountAPI.ts [hunk 9]
```diff
@@ -182,21 +184,23 @@ export class SmartAccountAPI extends BaseAccountAPI {
...
    */
-  async createUnsignedUserOp(info: TransactionDetailsForBatchUserOp): Promise<UserOperation> {
+  async createUnsignedUserOp(info: TransactionDetailsForBatchUserOp, vGasLimit?: BigNumber): Promise<UserOperation> {
     const { callData, callGasLimit } = await this.encodeUserOpCallDataAndGasLimit(info)
     console.log(callData, callGasLimit)
 
     const initCode = await this.getInitCode()
 
-    const verificationGasLimit = BigNumber.from(await this.getVerificationGasLimit())
+    let verificationGasLimit = vGasLimit
```
Comment (livingrockrises) [inline packages/account-abstraction/src/SmartAccountAPI.ts:193]: is there a reason to pass custom verificationGasLimit override it from outside? who will pass this, are we expecting devs to pass?

File: packages/account-abstraction/src/SmartAccountAPI.ts [hunk 10]
```diff
@@ -182,21 +184,23 @@ export class SmartAccountAPI extends BaseAccountAPI {
...
 
     let { maxFeePerGas, maxPriorityFeePerGas } = info
     // Note: Custom should be equal if it's for non EIP1559
+    const chainId = this.clientConfig.chainId
 
     if (maxFeePerGas == null || maxPriorityFeePerGas == null) {
-      const feeData = await this.provider.getFeeData()
-      Logger.log('EIP1559 feeData', feeData)
-      const chainId = this.clientConfig.chainId
+      const feeData = await this.httpRpcClient.getUserOpGasPrices(chainId)
```
Comment (livingrockrises) [inline packages/account-abstraction/src/SmartAccountAPI.ts:202]: let's have proper error handling here, if RPC url call to eth_getUserOpGasPrices then it should log and fallback to getting these values from provider like we're doing now.

File: packages/core-types/src/BundlerTypes.ts [hunk 11]
```diff
@@ -0,0 +1,7 @@
+export type UserOpGasPrices = {
+    status: number
+    message: string
+    gasPrice: string
+    maxFeePerGas: string
```
Comment (livingrockrises) [inline packages/core-types/src/BundlerTypes.ts:5]: do you think string is the right type to return?

File: packages/node-client/src/NodeClient.ts [hunk 12]
```diff
@@ -165,6 +165,10 @@ class NodeClient implements INodeClient {
   estimateHandlePaymentGas(
     estimateHandlePaymentTxGasDto: EstimateHandlePaymentTxGasDto
   ): Promise<EstimateGasResponse> {
+    const byPass = process.env.BY_PASS_ESTIMATION
```
Comment (livingrockrises) [inline packages/node-client/src/NodeClient.ts:168]: so this is at env level and not at dto level?
also BY_PASS naming convention looks odd

File: packages/node-client/src/NodeClient.ts [hunk 13]
```diff
@@ -185,6 +189,10 @@ class NodeClient implements INodeClient {
   async estimateHandlePaymentGasOverride(
     estimateHandlePaymentTxGasDto: EstimateHandlePaymentTxGasDto
   ): Promise<EstimateGasResponse> {
+    const byPass = process.env.BY_PASS_ESTIMATION
```
Comment (livingrockrises) [inline packages/node-client/src/NodeClient.ts:192]: better word for this could be 'skip' instead of by_pass

File: packages/node-client/src/NodeClient.ts [hunk 14]
```diff
@@ -185,6 +189,10 @@ class NodeClient implements INodeClient {
...
+    if ( byPass === 'true' ){
```
Comment (livingrockrises) [inline packages/node-client/src/NodeClient.ts:193]: and then this var also changes skipGasEstimation or skipHandlerPaymentGasEstimation

File: packages/smart-account/src/SmartAccount.ts [hunk 15]
```diff
@@ -587,18 +590,19 @@ class SmartAccount extends EventEmitter {
     if (updateImplTrx.data != '0x') {
       transactions.unshift(updateImplTrx)
     }
-    const response = await aaSigner.sendTransactionBatch(transactions, false)
+    const response = await aaSigner.sendTransactionBatch(transactions, transactionBatchDto.vGasLimit, transactionBatchDto.gasLimit, false)
```
Comment (livingrockrises) [inline packages/smart-account/src/SmartAccount.ts:593]: for batch you are sending verificationGasLimit and gasLimit as custom but not for regular sendTransaction looks like its just verification gasLimit

**Review by livingrockrises — COMMENTED** [review]
reviewed. ensure proper tests and error handling is in place.
changes should be well thought so they are aligned with what we'r planning to do with making sdk modular..
